src/pkg/scan/dao/scan/vulnerability.go

```diff
@@ -0,0 +1,231 @@
+	InsertForReport(ctx context.Context, reportUUID string, vr *VulnerabilityRecord) (int64, error)
+	// GetForReport gets vulnerability records for a report
+	GetForReport(ctx context.Context, reportUUID string) ([]*VulnerabilityRecord, error)
+	// GetForScanner gets vulnerability records for a scanner
+	GetForScanner(ctx context.Context, registrationUUID string) ([]*VulnerabilityRecord, error)
+	// DeleteForScanner deletes vulnerability records for a scanner
+	DeleteForScanner(ctx context.Context, registrationUUID string) (int64, error)
+	// DeleteForReport deletes vulnerability records for a report
+	DeleteForReport(ctx context.Context, reportUUID string) (int64, error)
+	// DeleteForDigests deletes vulnerability records for a provided list of digests
+	DeleteForDigests(ctx context.Context, digests ...string) (int64, error)
+	// GetRecordIdsForScanner gets record ids of vulnerability records for a scanner
+	GetRecordIdsForScanner(ctx context.Context, registrationUUID string) ([]int, error)
+
+}
+
+// NewVulnerabilityRecordDao returns a new dao to handle vulnerability data
+func NewVulnerabilityRecordDao() VulnerabilityRecordDao {
+	return &vulnerabilityRecordDao{}
+}
+
+type vulnerabilityRecordDao struct{}
+
+// Create creates new vulnerability record.
+func (v *vulnerabilityRecordDao) Create(ctx context.Context, vr *VulnerabilityRecord) (int64, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	_, vrID, err := o.ReadOrCreate(vr, "CVEID", "RegistrationUUID", "Digest", "Package", "PackageVersion")
+
+	return vrID, err
+}
+
+// Delete  deletes a vulnerability record
+func (v *vulnerabilityRecordDao) Delete(ctx context.Context, vr *VulnerabilityRecord) error {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+	_, err = o.Delete(vr, "CVEID", "RegistrationUUID")
+
+	return err
+}
+
+// List lists the vulnerability records with given query parameters.
+// Keywords in query here will be enforced with `exact` way.
+// If the registration ID (which = the scanner ID is not specified), the results
+// would contain duplicate records for a CVE depending upon the number of registered
+// scanners which individually store data about the CVE. In such cases, it is the
+// responsibility of the calling code to de-duplicate the CVE records or bucket them
+// per registered scanner
+func (v *vulnerabilityRecordDao) List(ctx context.Context, query *q.Query) ([]*VulnerabilityRecord, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	qt := o.QueryTable(new(VulnerabilityRecord))
+
+	if query != nil {
+		if len(query.Keywords) > 0 {
+			for k, v := range query.Keywords {
+				if vv, ok := v.([]interface{}); ok {
+					qt = qt.Filter(fmt.Sprintf("%s__in", k), vv...)
+					continue
+				}
+
+				qt = qt.Filter(k, v)
+			}
+		}
+
+		if query.PageNumber > 0 && query.PageSize > 0 {
+			qt = qt.Limit(query.PageSize, (query.PageNumber-1)*query.PageSize)
+		}
+	}
```

Comment:
These lines can be changed to `qs, err := orm.QuerySetter(ctx, query)`.

Reply:
I did not understand this comment. Also, I am not well versed with all Beego utilisation mechanisms. I would address this later if that is OK

Reply:
Sure.